src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5453,7 +5453,7 @@ static const int64_t right_3_bits = right_n_bits(3);
     }
     __ mv(c_rarg0, xthread);
     BLOCK_COMMENT("call runtime_entry");
-    __ call(runtime_entry);
+    __ rt_call(runtime_entry);
```

Comment:
I agree it's better to use `call_VM_leaf` for the Shenandoah cases. Then what about the changes in this file and templateInterpreterGenerator_riscv.cpp? Any reason to switch to `rt_call`?

Reply:
Old call():
```
    int32_t offset = 0;
    mv(temp, dest, offset); // =>li();
    jalr(x1, temp, offset);

```
To keep the sites the same (for non-code-cache calls)
New rt_call():
```
      movptr(tmp, target.target(), offset);
      Assembler::jalr(x1, tmp, offset);
```

Same here means absolute calls, no reloc required.
So I have tried to keep the calls the same. As you as we can optimize this by using reloc + la().

Reply:
Let me try to understand what you mean. So, are we going to remote the `relocate` call for non-code-cache call at [1] and further improve the `movptr` at [2] making use of `la`? This sounds interesting to me.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L5031
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L5033

Reply:
I have not looked at it in detail.

As mnemonic for **call** is _auipc_ + _jalr_, in hotspot `la()` + `jalr()`.
So sites using call() for non-code-cache was changed to rt_call(), which gets us the same result as the old call().

Hence this patch 'tries' to keep the generate assembly the same.

Reply:
Make sense?

Reply:
I am still think about the possibility of unifying `call` and `rt_call`. Having both of them could be confusing to me (and new comers I guess). Like this add-on change:
[addon.diff.txt](https://github.com/openjdk/jdk/files/15164874/addon.diff.txt)
What do you think?

Reply:
Off today, I'll have a look tomorrow, thanks.

Reply:
Yes, we should use is_32bit_offset_from_codecache to use auipc as much as possible.
As there is much churn in this patch and our testing takes so long I was trying to keep assembly same to avoid bisecting down issues.

It's much better with just two calls, either rt call or VM leaf.

I'll test your patch, thanks.

Reply:
We still need relocates rt_call, not sure why you removed it.
It seem like we need two version of rt_call one with address and one with Address.
Then it seem like we could remove far_call as the rt_call would do the right thing.

I like your idea, and we should do that, but it seems like it's not trivial just to add to this patch.
Is there a reason we need to include such in changes in this PR?

Reply:
> We still need relocates rt_call, not sure why you removed it.

I removed the relocate because I am thinking it should be absolute calls in the else block of `MacroAssembler::rt_call` [1], so no reloc required, as you mentioned in your previous comments. 

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L5027

> It seem like we need two version of rt_call one with address and one with Address. Then it seem like we could remove far_call as the rt_call would do the right thing.
> 
> I like your idea, and we should do that, but it seems like it's not trivial just to add to this patch. Is there a reason we need to include such in changes in this PR?

On holiday this week, we can discuss further next week :-)

Reply:
> We still need relocates rt_call, not sure why you removed it. It seem like we need two version of rt_call one with address and one with Address. Then it seem like we could remove far_call as the rt_call would do the right thing.

I guess maybe it's easier to only unify `call` and `rt_call` as a first step? And I prefer to keep `far_call` and `far_jump` as I think they are kind of different from rt_call. These two only handle the case when target is within code cache [1][2] as compared to rt_call which handles both code-cache and non-code-cache targets.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L3173
[2] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/macroAssembler_riscv.cpp#L3188

> I like your idea, and we should do that, but it seems like it's not trivial just to add to this patch. Is there a reason we need to include such in changes in this PR?

Another concern which also leads me to consider unifying `call` and `rt_call` is possible performance issues. I see some `call` are changed into `rt_call` like in file stubGenerator_riscv.cpp and templateInterpreterGenerator_riscv.cpp. As I mentioned in my previous comments, `rt_call` would emit a fixed-size `movptr` sequence (6 uncompressed instructions) for these call sites which invokes some C++ VM functions. But it's still possible for the original `call` to emit a more simpler auipc + jalr depending on the `is_32bit_offset_from_codecache` check in `la`, right? Hopefully, we can get rid of this issue with my proposed change.

Reply:
Yes, sorry I didn't explain that:
The cases in stubGenerator are all slow-path so we don't care much about performance.
Calling:
- MacroAssembler::debug64 debug
- JavaThread::check_special_condition_for_native_trans when suspend flag is set
- SharedRuntime::reguard_yellow_pages regaurd pages => syscall
- Interpreter::trace_code(t->tos_in()) Calling a interpreter trace method for the bytcode

Reply:
All right. What about the one in `MacroAssembler::call_VM_leaf_base`? The `call` there is replaced with movptr+jalr. Why not use `rt_call` here like the other places? Then we won't lose it when we eliminate `call` finally.

Reply:
Yes, you right the old:
call() = mv(tmp, adr) => li(adr) + jalr()

li(adr) is better than movptr()

I'll fix thanks!